opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -0,0 +1,249 @@
+				++numZeros;
+			}
+		}
+	}
+	return numZeros;
+}
+
+
+// convert matrix to blocked csr (bsr) arrays
+// if only_vals, do not convert rowPointers and colIndices
+// sparsity pattern should stay the same due to matrix-add-well-contributions
+template <class BridgeMatrix>
+void convertMatrixBsr(BridgeMatrix& mat, std::vector<double> &h_vals, std::vector<int> &h_rows, std::vector<int> &h_cols, int dim, bool only_vals) {
+	int sum_nnzs = 0;
+	int nnz = mat.nonzeroes()*dim*dim;
+
+	// copy nonzeros
+	memcpy(h_vals.data(), &(mat[0][0][0][0]), sizeof(double)*nnz);
+
+	// convert colIndices and rowPointers
+	if(only_vals == false){
+		h_rows.emplace_back(0);
+			for(typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r){
+				int size_row = 0;
+				for(auto c = r->begin(); c != r->end(); ++c){
+					h_cols.emplace_back(c.index());
+					size_row++;
+				}
+				sum_nnzs += size_row;
+				h_rows.emplace_back(sum_nnzs);
+			}
+		// set last rowpointer
+		h_rows[mat.N()] = mat.nonzeroes();
+	}
+} // end convertMatrixBsr()
+
+// converts the BlockVector b to a flat array
+template <class BridgeVector>
+void convertBlockVectorToArray(BridgeVector& b, std::vector<double> &h_b) {
+	memcpy(h_b.data(), &(b[0]), sizeof(double) * b.N() * b[0].dim());
+}
+#endif
+
+template <class BridgeMatrix, class BridgeVector>
+void BdaBridge::solve_system(BridgeMatrix *mat, BridgeVector &b, InverseOperatorResult &res)
+{
+
+#if HAVE_CUDA
+	if(use_gpu){
+		BdaResult result;
+		result.converged = false;
+		static std::vector<double> h_vals;
+		static std::vector<double> h_b;
+		static std::vector<int> h_rows;
+		static std::vector<int> h_cols;
+		int dim = (*mat)[0][0].N();
+		int N = mat->N()*dim;
+		int nnz = mat->nonzeroes()*dim*dim;
+
+		if(dim != 3){
+			std::cerr << "Error can only use cusparseSolver with blocksize = 3 at this time" << std::endl;
+			exit(1);
+		}
+
+		if(h_vals.capacity() == 0){
+			h_vals.reserve(nnz);
+			h_vals.resize(nnz);
+			h_b.reserve(N);
+			h_b.resize(N);
+			h_rows.reserve(N+1);
+			h_cols.reserve(nnz);
+		}
+
+#if PRINT_TIMERS_BRIDGE
+		Dune::Timer t_zeros;
+		int numZeros = checkZeroDiagonal(*mat);
+		printf("Checking zeros took %f s, found %d zeros\n", t_zeros.stop(), numZeros);
+#else
+		checkZeroDiagonal(*mat);
+#endif
+
+		bool initialized = backend->isInitialized();
+
+#if PRINT_TIMERS_BRIDGE
+		Dune::Timer t;
+#endif
+
+		convertMatrixBsr(*mat, h_vals, h_rows, h_cols, dim, initialized);
+		convertBlockVectorToArray(b, h_b);
+
+#if PRINT_TIMERS_BRIDGE
+		printf("Conversion to flat arrays: %.4f s\n", t.stop());
+#endif
+
+		/////////////////////////
+		// actually solve
+
+		if(initialized == false){
+			backend->initialize(N, nnz, dim);
+			backend->copy_system_to_gpu(h_vals.data(), h_rows.data(), h_cols.data(), h_b.data());
+			backend->analyse_matrix();
+		}else{
+			backend->update_system_on_gpu(h_vals.data(), h_b.data());
+		}
+		backend->reset_prec_on_gpu();
+		if(backend->create_preconditioner()){
```

Comment:
The function `cusparseXbsrilu02_zeroPivot` contains an implicit synchronization point, as it calls `cudaDeviceSynchronize`. But since `cusparseDbsrilu02` is a cusparse function, and the stream from the cusparseSolver is attached to the cusparse library, it should wait for the memcpys to be done.
There is a warning inside `create_preconditioner` if the preconditioner cannot be created.

Reply:
cusparseSolver now returns a cusparseSolverStatus value, which is interpreted by BdaBridge. The Bridge prints a warning using OpmLog.